### PR TITLE
KAFKA-12639: exit upon expired timer to prevent tight looping

### PR DESCRIPTION
**Pull request by philipnee**
https://issues.apache.org/jira/browse/KAFKA-12639

In AbstractCoordinator#joinGroupIfNeeded - joinGroup request will be retried without proper backoff, due to the expired timer. This is an uncommon scenario and possibly only appears during the testing, but I think it makes sense to enforce the client to drive the join group via poll.

### CODE REVIEW

**Comment by philipnee** [comment]
@guozhangwang - would you have time to review this 🥺 ?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java [hunk 1]
```diff
@@ -501,13 +501,16 @@ boolean joinGroupIfNeeded(final Timer timer) {
...
-                    exception instanceof RebalanceInProgressException ||
-                    exception instanceof MemberIdRequiredException)
+                        exception instanceof IllegalGenerationException ||
+                        exception instanceof RebalanceInProgressException ||
+                        exception instanceof MemberIdRequiredException)
                     continue;
                 else if (!future.isRetriable())
                     throw exception;
 
+                if (timer.isExpired()) {
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:513]: Could you add a couple comments here explaining why we check the timer again here in addition to in line 452 above? Maybe something like this:

```
We check the timer again after calling poll with the timer since it's possible that even after the timer has elapsed, the next client.poll(timer) would immediately return an error response which would cause us to not exiting the while loop.
```

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java [hunk 2]
```diff
@@ -501,13 +501,16 @@ boolean joinGroupIfNeeded(final Timer timer) {
                 }
 
                 if (exception instanceof UnknownMemberIdException ||
-                    exception instanceof IllegalGenerationException ||
-                    exception instanceof RebalanceInProgressException ||
-                    exception instanceof MemberIdRequiredException)
+                        exception instanceof IllegalGenerationException ||
+                        exception instanceof RebalanceInProgressException ||
+                        exception instanceof MemberIdRequiredException)
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:506]: Should we actually do the timer check before this? Since otherwise if the exception from the immediately returned responses is any of those four, we would still `continue` and skip the check below.

More concretely I think we can just move the remaining logic inside the `if` call:

```
if (!future.isRetriable()) {
    throw ..
} else {
    if (timer.isExpired()) {
        return false;
    } else if (exception instance of..) {
        continue;
    } else {
        timer.sleep(..)
    }
}
```

**Review by guozhangwang — COMMENTED** [review]
Thanks @philipnee . I left some comments.

Also I think we should add unit test with exactly the mocking clients  to 1) return a non-retriable exception, and check that we throw immediately (if we already have it in the test, then we can skip), 2) return one of the four exceptions, and check that we never sleep backoffs, 3) return other retriable exception, and check that we would still the remaining timer; and in case 2/3), we check that if timer has elapsed, we would also return false immediately.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java (hunk 2 above)
Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:506]: agreed. A comment here, retriableException check should happen after the `instanceOf` checks, because I think we actually want to retry upon these (according to the logic).

**Comment by philipnee** [comment]
Thanks @guozhangwang for the feedback - Added some tests there to cover the untesed cases.  I still have a quick question around this block, is it intentional to continue w/o sleep on the backoff timer? (quoting the original code)

```
if (exception instanceof UnknownMemberIdException ||
                    exception instanceof IllegalGenerationException ||
                    exception instanceof RebalanceInProgressException ||
                    exception instanceof MemberIdRequiredException)
                    continue;
                else if (!future.isRetriable())
                    throw exception;

                timer.sleep(rebalanceConfig.retryBackoffMs);
```

**Comment by guozhangwang** [comment]
> is it intentional to continue w/o sleep on the backoff timer?

Yes that's intentional. For those four exceptions, we'd like to send the follow-up request right away since the broker is waiting for those join-group request. But the question is, when the timer has already elapsed, should we honor that or should we ignore but always try to complete this mid-stage.

Since in the new protocol we would no longer have such mid-stages during a prepare_rebalance phase (cc @dajac to chime in if you feel different), I would suggest we respect the timer still for now to have a stronger `poll(timer)` timing guarantees.

**Comment by philipnee** [comment]
Thanks, @guozhangwang, that's my understanding as well.

**Comment by philipnee** [comment]
Moving the time check just broke a bunch of unit test 😅

**Comment by guozhangwang** [comment]
@philipnee is this the final version of this PR? Seems we are still honoring the four exceptions indicating the mid-stage of a rebalance more than the elapsed timer here?

**Comment by philipnee** [comment]
Hey @guozhangwang it's WIP - I think moving the timer check before the exception handling block (that 4 exceptions), kind of breaks a bunch of tests, as most tests are expecting the complete within a single poll. I'm looking into these breakage actually. sorry about the confusion.

**Comment by guozhangwang** [comment]
Oh got it, thanks! All good :) Please let me know when it's ready for a final look.

**Comment by philipnee** [comment]
Although, are we ok with handling these 4 exceptions the same way as before? I know you previously mentioned that it might be better off to make the rules more consistent, and I kind of agree with it.

**Comment by guozhangwang** [comment]
Yeah I think it's okay to make the rule consistent, i.e. to honor the timeout even under those four exceptions: if the timer has elapsed, then we should well return from the loop in

```
client.poll(future, timer);
            if (!future.isDone()) {
                // we ran out of time
                return false;
            }
```

even if the response yet to be returned would contain any of these four exceptions. So I think we should still obey this rule, i.e. even if a response has been returned and we know it's going to be one of these four exceptions, if the timer has elapsed, we still exit the loop.

**Comment by philipnee** [comment]
Hmm, strangely, this branch seems to trigger a bunch of initializing error failures. And I can't seem to reproduce them locally...

**Comment by philipnee** [comment]
Just a bit a note here on this PR: Seems like we need to be more deliberate at handling the timeout, because the non-retriable errors are always expected to be thrown. (except for the 4 cases), which is why the change triggered 60-ish breaking tests. Updating the PR to retrigger the test.

**Comment by philipnee** [comment]
The failures seem irrelevant to the change here: i.e. they dont' show up in both rounds.
```
Build / JDK 11 and Scala 2.13 / testDynamicListenerConnectionCreationRateQuota() – kafka.network.DynamicConnectionQuotaTest
41s
Build / JDK 17 and Scala 2.13 / testTaskRequestWithOldStartMsGetsUpdated() – org.apache.kafka.trogdor.coordinator.CoordinatorTest
2m 0s
Build / JDK 8 and Scala 2.12 / [1] Type=ZK, Name=testRegisterZkBrokerInKraft, MetadataVersion=3.4-IV0, Security=PLAINTEXT – kafka.server.KafkaServerKRaftRegistrationTest
7s
Build / JDK 8 and Scala 2.12 / [1] Type=ZK, Name=testRegisterZkBrokerInKraft, MetadataVersion=3.4-IV0, Security=PLAINTEXT – kafka.server.KafkaServerKRaftRegistrationTest
12s
```

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java [hunk 3]
```diff
@@ -500,14 +500,22 @@ boolean joinGroupIfNeeded(final Timer timer) {
                     requestRejoin(shortReason, fullReason);
                 }
 
+                // continue to retry as long as the timer hasn't expired
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: Could we simplify this multi-if logic as:

```
if (!future.isRetriable()) { throw }
else {
    if (timer.isExpired() { return false }
    else if (exception instance of.. ) { continue}
    else {timer.sleep(..)}
}
```

Also could we add a comment on top clarifying that the order of precedence are deliberated in this order and future changes should pay attention to not change it unnecessarily.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java [hunk 4]
```diff
@@ -1484,6 +1484,8 @@ public void testRebalanceWithMetadataChange() {
                 Utils.mkMap(Utils.mkEntry(topic1, 1), Utils.mkEntry(topic2, 1))));
         client.respond(joinGroupFollowerResponse(1, consumerId, "leader", Errors.NOT_COORDINATOR));
         client.prepareResponse(groupCoordinatorResponse(node, Errors.NONE));
+        coordinator.poll(time.timer(0)); // failing joinGroup request will require re-poll in order to retry
```
Comment (guozhangwang) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java]: It's not very clear to me why here and line 3403 below we need additional polls since the test scenarios seems irrelevant to error cases?

**Review by guozhangwang — COMMENTED** [review]
@philipnee thanks for the added unit test. I made another pass and left some more comments.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java (hunk 3 above)
Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: I think the `instanceof ...` exceptions are also non-retriable, and I think they need to be handled first.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java (hunk 4 above)
Comment (philipnee) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java:1487]: good point... let me update the tests there. thanks

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java (hunk 3 above)
Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: so the if else blocks becomes a bit fragmented.  or we could do:
```
if (!future.isRetriable()) {
  if ( ... instance of ... ) { continue; }
  throw ...
}

{rest of the logic there}
```

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java (hunk 4 above)
Comment (philipnee) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java]: The NOT_COORDINATOR error originally should trigger retries; however, in the new code, it would exit due to an expired timer. Another way to do it is using poll(time.timer(1))

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java [hunk 5]
```diff
@@ -3398,7 +3400,8 @@ public void testPrepareJoinAndRejoinAfterFailedRebalance() {
             client.respond(syncGroupResponse(partitions, Errors.NONE));
 
             // Join future should succeed but generation already cleared so result of join is false.
-            res = coordinator.joinGroupIfNeeded(time.timer(1));
+            coordinator.joinGroupIfNeeded(time.timer(0));
```
Comment (philipnee) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java]: Similar here, the timer is expired upon IllegalGenerationException, the loop would continued in the original code, but now it would exit.  I guess we could try to poll for a bit longer, like 3ms instead of 0ms.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java (hunk 3 above)
Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: However, this is a bit more nested, which can be harder to read

**Comment by philipnee** [comment]
Here are a couple things I updated:
1. Added some documentation to clarify the intent, but I didn't rewrite it as nested if "can be" harder to read.
2.  Added non zero timeouts for the tests as our timer now is stricter and will explicitly exit upon expiration.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java (hunk 4 above)
Comment (guozhangwang) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java]: Ack, that makes sense.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java [hunk 6]
```diff
@@ -500,14 +500,24 @@ boolean joinGroupIfNeeded(final Timer timer) {
                     requestRejoin(shortReason, fullReason);
                 }
 
+                // 4 special non-retriable exceptions that we want to retry, as long as the timer hasn't expired.
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: I think here the comment is not to just state what the code did, since readers can just understand that from the code :P instead what we want to emphasize is to remind future contributors that they should be careful to not change the precedence ordering of this logic unnecessarily.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java [hunk 7]
```diff
@@ -500,14 +500,24 @@ boolean joinGroupIfNeeded(final Timer timer) {
...
                 if (exception instanceof UnknownMemberIdException ||
-                    exception instanceof IllegalGenerationException ||
-                    exception instanceof RebalanceInProgressException ||
-                    exception instanceof MemberIdRequiredException)
+                        exception instanceof IllegalGenerationException ||
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:504]: Ah thanks for the clarifications!

Thinking about this a bit more (sorry for getting back and forth..), I now concerned a bit more that for some usage patterns where `poll` call would be triggered less frequently, we may not be coming back to handle these four exceptions while at the same time the broker is ticking and waiting for the join-group request to be re-sent. Hence I'm changing my mind to lean a bit more to honor the exception types for immediate handling than the timeouts --- again, sorry for going back and forth...

So I think we would define the ordering as the following:

1. For un-retriable exception, always try to handle immediately and not honor the timer.
2. Otherwise, honor the timer.

In that case, we could just go back to the first time you made the change, i.e. just add the 

```
if (timer.isExpired())
                        return false;
```

After the `if/else-if` block. Still it's better to comment that above ordering is diligently designed as such.

Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:504]: Hey thanks for the comments again and absolutely no apology is needed there! I guess, as we all know, rebalancing is full of subtleties, so it makes sense to be careful about these non-retriable exception case. I think it's a good idea to keep the original behavior consistent, in case of unexpected breakage. Updating the PR.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java [hunk 8]
```diff
@@ -1484,7 +1484,8 @@ public void testRebalanceWithMetadataChange() {
                 Utils.mkMap(Utils.mkEntry(topic1, 1), Utils.mkEntry(topic2, 1))));
         client.respond(joinGroupFollowerResponse(1, consumerId, "leader", Errors.NOT_COORDINATOR));
         client.prepareResponse(groupCoordinatorResponse(node, Errors.NONE));
-        coordinator.poll(time.timer(0));
+        assertFalse(client.hasInFlightRequests());
+        coordinator.poll(time.timer(1));
```
Comment (philipnee) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java:1488]: note: we need to add a timeout here to give the retry a second chance, because in the new code, the timer is checked and causes the method to exit.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java [hunk 9]
```diff
@@ -501,13 +501,18 @@ boolean joinGroupIfNeeded(final Timer timer) {
...
 
                 if (exception instanceof UnknownMemberIdException ||
-                    exception instanceof IllegalGenerationException ||
-                    exception instanceof RebalanceInProgressException ||
-                    exception instanceof MemberIdRequiredException)
+                        exception instanceof IllegalGenerationException ||
+                        exception instanceof RebalanceInProgressException ||
+                        exception instanceof MemberIdRequiredException)
                     continue;
                 else if (!future.isRetriable())
```
Comment (philipnee) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:508]: the previous logic was reverted with some autocorrection to the indentation.

**Review by guozhangwang — APPROVED** [review]
LGTM, waiting for the jenkins job to complete.

**Comment by philipnee** [comment]
Hmm. I think these tests are flaky actually
```
Build / JDK 17 and Scala 2.13 / shouldPauseStandbyTaskAndNotTransitToUpdateStandbyAgain() – org.apache.kafka.streams.processor.internals.DefaultStateUpdaterTest
30s
Build / JDK 17 and Scala 2.13 / shouldPauseActiveTaskAndTransitToUpdateStandby() – org.apache.kafka.streams.processor.internals.DefaultStateUpdaterTest
30s
Build / JDK 17 and Scala 2.13 / testTaskRequestWithOldStartMsGetsUpdated() – org.apache.kafka.trogdor.coordinator.CoordinatorTest
2m 0s
Build / JDK 11 and Scala 2.13 / testListenerConnectionRateLimitWhenActualRateAboveLimit() – kafka.network.ConnectionQuotasTest
19s
Build / JDK 11 and Scala 2.13 / shouldRemovePausedAndUpdatingTasksOnShutdown() – org.apache.kafka.streams.processor.internals.DefaultStateUpdaterTest
30s
Build / JDK 11 and Scala 2.13 / shouldPauseStandbyTaskAndNotTransitToUpdateStandbyAgain() – org.apache.kafka.streams.processor.internals.DefaultStateUpdaterTest
31s
```

**Comment by guozhangwang** [comment]
The test failures are not relevant (but some of them are related to DefaultStateUpdaterTest.. sigh).

**Comment by guozhangwang** [comment]
Merged to trunk.

**Comment by philipnee** [comment]
yeah the DefaultStateUpdaterTest has been failing from time to time... not sure why 😭
